src/coreclr/jit/gentree.cpp

```diff
@@ -14881,6 +14881,21 @@ GenTree* Compiler::gtFoldExprConst(GenTree* tree)
                                 goto CNS_LONG;
 
                             case TYP_FLOAT:
+                            {
+#if defined(TARGET_64BIT)
+                                if (tree->IsUnsigned())
+                                {
+                                    f1 = (float)UINT32(i1);
```

Comment:
Is the problem we were converting UINT32 to double when calling `forceCastToFloat`, before converting to `float` in that function?

Seems like we should get rid of `forceCastToFloat` and `forceCastToUInt32` (their comments refer to ancient build compilers). But maybe all casts to/from floats should be explicit function calls to `FloatingPointUtils` functions. E.g., here maybe add `FloatingPointUtils::convertUInt32ToFloat()`?

Reply:
Right, the problem was that it was doing `uint->double->float` which can produce a different result from `uint->float` in some edge cases.

64-bit currently does `uint->long->float`; where-as 32-bit currently does `uint->long->double->float`, so this ensures the constant folding matches that behavior.

Longer term it would be beneficial to use helpers for these conversions, but that is a much bigger and more involved change. I tried to keep this as minimal as possible for .NET 8 to avoid churn/risk.